[zhacatomn] iP

zhacatomn opened the pull request:
# Duke-Lite

> "I am a duck" - Duck

Duke-lite is **simple** and *easy to use*, with the following features:
- tagging
- that's basically it

It's very easy to get started 😄:
1. Download the jar file [here](https://github.com/zhacatomn/ip/releases/tag/A-Release)
2. Run it

Things to do:
- [X] Tagging
- [ ] More different types of tasks

Here's a small code snippet:
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Duke.class, args);
    }
}

```

Commit:
Add Gradle support

zhacatomn committed:
Complete level 1

zhacatomn committed:
Complete level 2

zhacatomn committed:
Complete level 3

zhacatomn committed:
Complete level 4

zhacatomn committed:
Implement I/O testing

zhacatomn committed:
Complete level 5

zhacatomn committed:
Complete level 6

zhacatomn committed:
Add javadocs

zhacatomn committed:
Organise imports

zhacatomn committed:
Wrap in pacakage

zhacatomn committed:
Fix minor issue

zhacatomn committed:
Complete level 7

zhacatomn committed:
Complete level 8

zhacatomn committed:
Merge branch 'branch-Level-8'

zhacatomn committed:
Merge branch 'add-gradle-support'

zhacatomn committed:
Refactor to use OOP

zhacatomn committed:
Complete level 9

zhacatomn committed:
Add javadocs

zhacatomn committed:
Modify exit message to be a constant variable

zhacatomn committed:
Add jUnit tests

zhacatomn committed:
Refactor into new package 'task'

LiuMC-SG commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * Returns a Duke object.
+     * @param filePath The file path where the data file is located.
+     */
```
I noticed that you perhaps forgot the empty line between description and parameter sections. I noticed the same issue in several other places too. Perhaps you should use the shortcut "Ctrl+Alt+L" in Intellij for autoformatting.

LiuMC-SG commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,57 @@
+    static protected final String DATA_DIR = "data/";
+    static protected final String DATA_FILENAME = "duke.txt";
+    static protected final String WELCOME_MESSAGE = "Hello! I'm Duke\n\tWhat can I do for you?";
+    static protected final String EXIT_MESSAGE = "Bye. Hope to see you again soon!";
```
I like that you seperated out the strings to be used.

LiuMC-SG commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+    public void parseString(String inputStr, TaskList tasks, Ui ui) throws BadCommandException {
+        inputStr = inputStr.trim();
+        try {
+            if (inputStr.equals(Command.LIST.toString())) {
```
Maybe you can consider switch-case statements?

LiuMC-SG commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+                String[] paramsSplit = params.split("/from|/to", 3);
+                if (paramsSplit.length < 3) {
+                    throw new BadCommandException("There are insufficient parameters!");
+                }
```
I like the fact that you use split based on 2 parameters. :D

LiuMC-SG commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,123 @@
+import java.util.Collection;
+
+public class TaskList {
+    final static BadCommandException INDEX_OOB_ERROR =
+            new BadCommandException("Index given is out of bounds!");
```
Maybe this can be declared within badcommandexception class and be used as a static exception (i.e. Using `BadCommandException.INDEX_OOB_ERROR`)

LiuMC-SG commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,123 @@
+public class TaskList {
+    final static BadCommandException INDEX_OOB_ERROR =
+            new BadCommandException("Index given is out of bounds!");
+    protected ArrayList<Task> tasks;
```
I feel like this should be private in accordance to Java guidelines. There are multiple of these instances later on as well. Since the methods to modify the class instance variables should be done within class methods itself.

LiuMC-SG commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,123 @@
+        tasks = new ArrayList<>(initTasks);
+    }
+
+    private boolean isIndexInRange(int idx) {
+        return idx < tasks.size() && idx >= 0;
+    }
```
Seems like you missed out a JavaDoc here. There are some other instances as well.

LiuMC-SG commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,64 @@
+package duke.task;
+
+public abstract class Task {
+    public enum TaskIcon {
+        TODO ("T"),
+        EVENT ("E"),
+        DEADLINE ("D");
+        private final String symbol;
+        TaskIcon(String symbol) {
+            this.symbol = symbol;
+        }
```
Good use of enums here

LiuMC-SG reviewed (commented):
Good work so far. You seem to have all of the basics completed. Keep up the good work. :D

zhacatomn committed:
Complete level 10

zhacatomn committed:
Re-style UI

zhacatomn committed:
Refactor Parser

Parser uses nested conditionals and there's a lot of repeated code.

This makes readability difficult, and adding new commands troublesome.

Replacing the conditionals with a switch statement will allow the code to be more readable. Abstracting out many of the behaviors of the commands will also help to reduce code duplication.

Let's,
* Refactor the conditional statement into a switch statement using the Command enum
* Abstract out the behavior to extract delimiters from the parameter string into its own private function

zhacatomn committed:
Amend javadocs

zhacatomn committed:
Add assertions

zhacatomn committed:
Merge pull request #1 from zhacatomn/improve-code-quality

Improving code quality of the parser

zhacatomn committed:
Merge branch 'master' into add-assertions

zhacatomn committed:
Merge pull request #2 from zhacatomn/add-assertions

Add assertions

zhacatomn committed:
Refactor deadline and event cases for parser

QQH0828 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,200 @@
+            new BadCommandException("I'm sorry, but I don't know what that means :-(");
+    private static final String PARAMS_DELIMITER =  "/";
+
+    private enum Command {
+        LIST("list"),
+        MARK("mark"),
+        UNMARK("unmark"),
+        DELETE("delete"),
+        TODO("todo"),
+        DEADLINE("deadline"),
+        EVENT("event"),
+        FIND("find");
+        private final String commandStr;
+        private static final HashMap<String, Command> STRING_TO_COMMAND_MAP = new HashMap<>();
+        static {
+            for (Command command: values()) {
+                STRING_TO_COMMAND_MAP.put(command.commandStr, command);
+            }
+        }
```
I notice that this class feels a bit long when I am reading it, maybe can separate the enum Command into a new class.

QQH0828 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,200 @@
+        // then we take the first item in paramsSplit as the default argument.
+        int startingIdx = hasDefaultArgument ? 1 : 0;
+        String defaultArgument = hasDefaultArgument ? paramsSplit[0].trim() : null;
+
+        HashMap<String, String> paramToArgMap = new HashMap<>();
+        for (int i = startingIdx; i < paramsSplit.length; i++) {
```
Ooh, I like your clever usage of hashmap here!

QQH0828 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,129 @@
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+public class Storage {
```
I notice that you seem missing the heading of the class such as 
/**
    Storage class to ....
**/

QQH0828 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,123 @@
+        return matchingTasks;
+    }
+
+    public int getSize() {
+        return tasks.size();
+    }
```
I notice that seems like you missed out on a JavaDoc here.

QQH0828 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,64 @@
+        isDone = false;
+    }
+
+    public boolean isDone() {
+        return isDone;
+    }
+
+    public TaskIcon getTaskIcon() {
+        return this.taskIcon;
+    }
```
I notices that seems like you missed out on a JavaDoc at some short methods.

QQH0828 reviewed (commented):
LGTM! Overall, Most of the coding standards were adhered to! Except for a few places
where the class headings are missing and the java doc of the methods!

zhacatomn committed:
Update TaskList.java

zhacatomn committed:
Refactor out createNewTask in Parser

zhacatomn committed:
Add basic tagging functionality

zhacatomn committed:
Improve date detection for deadline and events

zhacatomn committed:
Add bye command

zhacatomn committed:
Add Ui.png

zhacatomn committed:
Update README.md

zhacatomn committed:
Fix list in docs

zhacatomn committed:
Finish command list in docs

zhacatomn committed:
Fix typos in User Guide

zhacatomn committed:
Update build.gradle

zhacatomn committed:
Update README.md

zhacatomn committed:
Fix checkstyle issues

zhacatomn committed:
Refactor to use fxml
